Libraries/src/Amazon.Lambda.ApplicationLoadBalancerIdentity/ALBIdentityMiddleware.cs

```diff
@@ -0,0 +1,245 @@
+        private const string AuthenticationType = "oidc";
+        private const string ContentTypeHeader = "Content-Type";
+        private const string ContentTypeValue = "text/plain";
+
+        /// Caching related properties
+        private static uint[] byteLookup;
+        private readonly bool cacheIdentities;
+        private readonly MemoryCache cachedIds;
+        private readonly SHA1 sha;
+        private readonly TimeSpan cacheDuration;
+
+        // Validation related properties
+        private readonly string region;
+        private readonly ConcurrentDictionary<string, Task<TokenValidationParameters>> validationParameters;
+        private readonly JsonWebTokenHandler tokenHandler;
+
+        // Properties from DI
+        private readonly ILogger<ALBIdentityMiddleware> logger;
+        private readonly ALBIdentityMiddlewareOptions options;
+        private readonly RequestDelegate next;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ALBIdentityMiddleware"/> class.
+        /// </summary>
+        /// <param name="next">The next <see cref="RequestDelegate"/> to call in the chain.</param>
+        /// <param name="logger"><see cref="ILogger{TCategoryName}"/> implementation.</param>
+        /// <param name="options">An instance of the <see cref="ALBIdentityMiddlewareOptions"/> class containing the configuration settings for the Middleware.</param>
+        public ALBIdentityMiddleware(RequestDelegate next, ILogger<ALBIdentityMiddleware> logger, ALBIdentityMiddlewareOptions options)
+        {
+            this.next = next;
+            this.logger = logger;
+            this.options = options;
+
+            if (this.options.VerifyTokenSignature)
+            {
+                if (InternalHttpClient == null) InternalHttpClient = new HttpClient();
+                this.region = Environment.GetEnvironmentVariable(AWSRegionEnvironmentVariable);
+                this.validationParameters = new ConcurrentDictionary<string, Task<TokenValidationParameters>>();
+                this.tokenHandler = new JsonWebTokenHandler();
+            }
+
+            this.cacheIdentities = this.options.MaxCacheSizeMB.HasValue;
+            if (this.cacheIdentities)
+            {
+                // TimeSpan.MaxValue isn't a real thing. If the underlying library that uses it converts it to an integer (which a lot do),
+                // then it will throw an ArgumentOutOfRange exception if you specify a timespan with more seconds in it than
+                // the max value of an Int32 has. To use an *actual* MaxValue that doesn't fail, you need to create a timespan from
+                // the maximum int size in seconds.
+                var maxCacheLife = TimeSpan.FromSeconds(int.MaxValue);
+
+                this.cachedIds = new MemoryCache(new MemoryCacheOptions
+                {
+                    // Evict 20% of cache entries on max size.
+                    CompactionPercentage = 0.2,
```

Comment:
Should this be configurable from the options object?

Reply:
Seems reasonable.